firedrake/assign.py

```diff
@@ -193,18 +220,15 @@ def _function_weights(self):
         return tuple(w for (c, w) in self._weighted_coefficients
                      if isinstance(c, Function))
 
-    @property
-    def _indices(self):
-        return self._subset.indices if self._subset else ...
-
-    # TODO: It would be more efficient in permissible cases to use VecMAXPY instead of numpy operations.
-    def _assign_single_dat(self, assignee_dat, function_dats):
-        assignee_dat.data_with_halos[self._indices] = self._compute_rvalue(function_dats)
+    def _assign_single_dat(self, lhs, indices, rvalue, assign_to_halos):
+        if assign_to_halos:
+            lhs.data_with_halos[indices] = rvalue
+        else:
+            lhs.data[indices] = rvalue
```

Comment:
`.data_with_halos` provokes a halo exchange on `lhs`. Do you want this, given you're overwriting things? i.e. does pyop2 need `.data_wo_with_halos` and `.data_wo`?

Reply:
Good catch! I agree that we should have `data_wo` and `data_wo_with_halos`. Will implement.

Reply:
Done.